sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/JsonUtilsTest.java

```diff
@@ -0,0 +1,308 @@
+      Arrays.asList(
+          testCase(Cat.of("Figaro", 12.3, 8, true)),
+          testCase(NestedCat.of(Cat.of("QuantumCat", -100.123, 999, false))),
+          testCase(
+              ArrayOfCats.of(
+                  Arrays.asList(
+                      Cat.of("Chill Cat", 1903.1514, 1, true),
+                      Cat.of("Toasty Cat", 5.9, 4, true),
+                      Cat.of("Hangry Cat", 3.14, 3, true)))));
+
+  @Test
+  public void testGetJsonBytesToRowFunction() {
+    for (TestCase<? extends RowEncodable> caze : testCases) {
+      Row expected = caze.row;
+      Row actual = JsonUtils.getJsonBytesToRowFunction(expected.getSchema()).apply(caze.jsonBytes);
+      assertEquals(caze.userT.toString(), expected, actual);
+    }
+  }
+
+  @Test
+  public void testGetJsonStringToRowFunction() {
+    for (TestCase<? extends RowEncodable> caze : testCases) {
+      Row expected = caze.row;
+      Row actual =
+          JsonUtils.getJsonStringToRowFunction(expected.getSchema()).apply(caze.jsonString);
+      assertEquals(caze.userT.toString(), expected, actual);
+    }
+  }
+
+  @Test
+  public void testGetRowToJsonBytesFunction() {
+    for (TestCase<? extends RowEncodable> caze : testCases) {
+      byte[] expected = caze.jsonBytes;
+      byte[] actual = JsonUtils.getRowToJsonBytesFunction(caze.row.getSchema()).apply(caze.row);
+      assertJsonEquals(caze.userT.toString(), expected, actual);
+    }
+  }
+
+  @Test
+  public void testGetRowToJsonStringsFunction() {
+    for (TestCase<? extends RowEncodable> caze : testCases) {
+      String expected = caze.jsonString;
+      String actual = JsonUtils.getRowToJsonStringsFunction(caze.row.getSchema()).apply(caze.row);
+      assertJsonEquals(caze.userT.toString(), expected, actual);
+    }
+  }
+
+  private static <T extends RowEncodable> TestCase<T> testCase(T cat) {
+    return new TestCase<>(cat);
+  }
+
+  private static class TestCase<T extends RowEncodable> {
+
+    final String name;
+    final T userT;
+    final String jsonString;
+    final byte[] jsonBytes;
+    final Row row;
+
+    private TestCase(T userT) {
+      this.name = userT.toString();
+      this.userT = userT;
+      this.jsonString = GSON.toJson(userT);
+      this.jsonBytes = JsonUtils.jsonStringToByteArray(jsonString);
+      this.row = userT.toRow();
+    }
+  }
+
+  @DefaultSchema(JavaBeanSchema.class)
+  private static class Cat implements RowEncodable {
+    private static final TypeDescriptor<Cat> TYPE_DESCRIPTOR = TypeDescriptor.of(Cat.class);
+    private static final SerializableFunction<Cat, Row> TO_ROW_FUNCTION =
+        JAVA_BEAN_SCHEMA.toRowFunction(TYPE_DESCRIPTOR);
+
+    static Cat of(String name, double weight, int age, boolean spayNeutered) {
+      return new Cat(name, weight, age, spayNeutered);
+    }
+
+    private final String name;
+    private final double weight;
+    private final int age;
+    private final boolean spayNeutered;
```

Comment:
I wonder if it makes sensse to have other types (float, long, DateTime)

Reply:
Good point!  I will add these in an additional PR.